query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,74 @@
+import { DirectiveNode, GraphQLCompositeType, GraphQLObjectType, ValueNode } from "graphql";
+
+export class Scope<TParent extends GraphQLCompositeType> {
+  private cachedIdentityKey: string | undefined = undefined;
+  public readonly possibleTypes: ReadonlyArray<GraphQLObjectType>;
+
+  constructor(
+    public readonly parentType: TParent,
+    possibleTypes: ReadonlyArray<GraphQLObjectType>,
+    public readonly directives?: ReadonlyArray<DirectiveNode>) {
+    // We need to sort the possible types to make our `identityKey()` method work.
+    this.possibleTypes = [...possibleTypes].sort();
+  }
+
+  private valueIdentityKey(value: ValueNode) : string {
```

Comment:
Just a quick drive-by comment, not reviewing whole PR.

What standard do these functions have to meet? Do we want to ensure that distinct values (and directives) always map to distinct keys? (I think the answer to this question should be in a comment — if it's fine to have clashes then say so.)

In that case I think we need a fair amount more complexity.  eg, the scalars (at least) should include their type so that `1` and `"1"` are different; lists should be able to differentiate between `["a-b", "c"]` and `["a", "b-c"]`, etc.

One possibility would be just to use the graphql-js printer (`print`), perhaps followed by `stripIgnoredCharacters` and/or a hash or something

Reply:
Good point. Switched to `print`/`stripIgnoredCharacters`.

Reply:
One may noticed that in the reworked branch I've just pushed, I've actually removed the use of `print`/`stripIgnoredCharacters` and reverted back to a `valueIdentityKey` method.

The reason is that we ideally want the equality to be a bit more semantic. In particular, it makes sense that argument be non meaningful. Meaning that `@foo(a:1, b:2)` be equal to `@foo(b:2, a:1)`. Doing so is easier by generating the string manually and sorting arguments (fields in input objects are also sorted similarly, which is why we don't use `print` for values either).

I did update the method to distinguish between of different types.

The whole business is arguably a bit hacky, but doing that more cleanly (through a proper equality method) is actually quite a bit more code in this case in javascript, and we agree with @martijn that this was probably not worth bothering.